crates/bevy_ui/src/render/gradient.rs

```diff
@@ -0,0 +1,984 @@
+                                f32::max(clip.min.x - positions[3].x, 0.),
+                                f32::min(clip.max.y - positions[3].y, 0.),
+                            ),
+                        ]
+                    } else {
+                        [Vec2::ZERO; 4]
+                    };
+
+                    let positions_clipped = [
+                        positions[0] + positions_diff[0].extend(0.),
+                        positions[1] + positions_diff[1].extend(0.),
+                        positions[2] + positions_diff[2].extend(0.),
+                        positions[3] + positions_diff[3].extend(0.),
+                    ];
+
+                    let points = [
+                        points[0] + positions_diff[0],
+                        points[1] + positions_diff[1],
+                        points[2] + positions_diff[2],
+                        points[3] + positions_diff[3],
+                    ];
+
+                    let transformed_rect_size = gradient.transform.transform_vector3(rect_size);
+
+                    // Don't try to cull nodes that have a rotation
+                    // In a rotation around the Z-axis, this value is 0.0 for an angle of 0.0 or π
+                    // In those two cases, the culling check can proceed normally as corners will be on
+                    // horizontal / vertical lines
+                    // For all other angles, bypass the culling check
+                    // This does not properly handles all rotations on all axis
+                    if gradient.transform.x_axis[1] == 0.0 {
+                        // Cull nodes that are completely clipped
+                        if positions_diff[0].x - positions_diff[1].x >= transformed_rect_size.x
+                            || positions_diff[1].y - positions_diff[2].y >= transformed_rect_size.y
+                        {
+                            continue;
+                        }
+                    }
+
+                    let uvs = { [Vec2::ZERO, Vec2::X, Vec2::ONE, Vec2::Y] };
+
+                    let mut flags = if gradient.node_type == NodeType::Border {
+                        shader_flags::BORDER
+                    } else {
+                        0
+                    };
+
+                    let (g_start, g_dir, g_flags) = match gradient.resolved_gradient {
+                        ResolvedGradient::Linear { angle } => {
+                            let angle = angle.rem_euclid(TAU);
+                            (
+                                points[if angle < TAU / 4. {
+                                    3
+                                } else if angle < TAU / 2. {
+                                    0
+                                } else if angle < 3. * TAU / 4. {
+                                    1
+                                } else {
+                                    2
+                                }]
+                                .into(),
```

Comment:
can probably be extracted to a function

Reply:
Oh this looks stupid, it doesn't need a branch here, just need to divide by four and round and shift the result